minimint-api/src/module/mod.rs

```diff
@@ -19,25 +20,94 @@ pub struct InputMeta<'a> {
     pub puk_keys: Box<dyn Iterator<Item = XOnlyPublicKey> + 'a>,
 }
 
-/// Map of URL parameters and their values.
-pub type Params<'a> = HashMap<&'static str, &'a str>;
+#[derive(Debug)]
+pub struct ApiError {
+    pub code: i32,
+    pub message: String,
+}
+
+impl ApiError {
+    pub fn new(code: i32, message: String) -> Self {
+        Self { code, message }
+    }
+
+    pub fn not_found(message: String) -> Self {
+        Self::new(404, message)
+    }
+
+    pub fn bad_request(message: String) -> Self {
+        Self::new(400, message)
+    }
+}
+
+#[async_trait]
+pub trait TypedApiEndpoint {
+    type State: Sync;
+
+    /// example: /transaction
+    const PATH: &'static str;
+
+    type Param: serde::de::DeserializeOwned + Send;
+    type Response: serde::Serialize;
+
+    async fn handle(state: &Self::State, params: Self::Param) -> Result<Self::Response, ApiError>;
+}
+
+#[macro_export]
+macro_rules! __api_endpoint {
+    (
+        $path:expr,
+        async |$state:ident: &$state_ty:ty, $param:ident: $param_ty:ty| -> $resp_ty:ty $body:block
+    ) => {{
+        struct Endpoint;
+
+        #[async_trait::async_trait]
+        impl $crate::module::TypedApiEndpoint for Endpoint {
+            const PATH: &'static str = $path;
+            type State = $state_ty;
+            type Param = $param_ty;
+            type Response = $resp_ty;
+
+            async fn handle(
+                $state: &Self::State,
+                $param: Self::Param,
+            ) -> ::std::result::Result<Self::Response, $crate::module::ApiError> {
+                $body
+            }
+        }
+
+        $crate::module::__create_api_endpoint::<Endpoint>()
+    }};
+}
+
+pub const fn __create_api_endpoint<E: TypedApiEndpoint>() -> ApiEndpoint<E::State> {
```

Comment:
for now, I have added bbac797
We can squash or remove it.

Reply:
I'm happy we are off nightly, but other than that I think it would be too limiting to have a MSRV, especially since the useful ones (packaged in debian) are always quite old (1.48).

Reply:
decided to just inline the function, it is not too bad for readability.

updating cargo in nix is much more annoying for me atleast.

Reply:
"MSRV enforced by difficulty of updating CI" wasn't what I was going for with the nix stuff :see_no_evil: